src/main/java/io/leonis/torch/component/CanvasComponent.java

```diff
@@ -0,0 +1,42 @@
+package io.leonis.torch.component;
+
+import com.googlecode.lanterna.*;
+import com.googlecode.lanterna.gui2.*;
+import java.util.function.BiFunction;
+import java.util.stream.IntStream;
+import lombok.AllArgsConstructor;
+
+/**
+ * The class CanvasComponent.
+ *
+ * Draws an rectangle using {@code backgroundSupplier} to determine the color at each coordinate
+ *
+ * @author Jeroen de Jong
+ */
+@AllArgsConstructor
+public class CanvasComponent extends AbstractComponent<CanvasComponent> {
+
+  /**
+   * Function to determine background color at a specific position
+   */
+  private final BiFunction<Integer, Integer, TextColor> backgroundSupplier;
+
+  @Override
+  protected ComponentRenderer<CanvasComponent> createDefaultRenderer() {
+    return new ComponentRenderer<CanvasComponent>() {
+      @Override
+      public TerminalSize getPreferredSize(final CanvasComponent component) {
+        return new TerminalSize(0, 0);
+      }
+
+      @Override
+      public void drawComponent(final TextGUIGraphics graphics, final CanvasComponent component) {
+        IntStream.range(0, this.getPreferredSize(component).getColumns()).forEach(x ->
+            IntStream.range(0, this.getPreferredSize(component).getRows()).forEach(y -> {
```

Comment:
This could be reduced to one `IntStream`; but i rather keep it the way it is.
```java
int width = this.getPreferredSize(component).getRows();
int height = this.getPreferredSize(component).getColumns();
return IntStream.range(0, height * width)
    .mapToObj(i -> new TerminalPosition(i % width, i / width));
```

Reply:
```java
IntStream.range(0, height)
    .flatMap(row -> 
        IntStream.range(0, width)
            .map(col -> new TerminalPosition(row, col)))
```

Reply:
Doesn't compile; `IntStream#flatMap` needs an int. you need `flatMapToObj` but that doesn't exist; so you'd need something like this:
```java
IntStream.range(0, height)
    .mapToObj(row -> 
        IntStream.range(0, width)
                 .mapToObj(col -> new TerminalPosition(row, col)))
    .flatMap(Function.identity())
```

Reply:
you can just put `IntStream.range(0, width).boxed()` and then you can use `map` properly

Reply:
neat; didn't know that. It would mean that the BiFunction needs to be changed; but that seems fine.

Reply:
The full snippet would become
```java
        return IntStream.range(0, this.getPreferredSize(component).getColumns()).boxed()
         .flatMap(row -> 
             IntStream.range(0, this.getPreferredSize(component).getRows())
                .mapToObj(col -> new TerminalPosition(row, col)))
          .forEach(position -> {
              graphics.setBackgroundColor(backgroundSupplier.apply(position);
              graphics.setCharacter(position, ' ');
          });
```

over
```java
        IntStream.range(0, this.getPreferredSize(component).getColumns()).forEach(x ->
            IntStream.range(0, t).forEach(y -> {
              graphics.setBackgroundColor(backgroundSupplier.apply(x, y));
              graphics.setCharacter(x, y, ' ');
            }));
```

I don't think it's an improvement.